src/ship/PlayerShipController.cpp

```diff
@@ -149,13 +289,224 @@ void PlayerShipController::PostLoadFixup(Space *space)
+				strafe += -m_ship->GetOrient().VectorX() * maxStrafeSpeed * Util::sign(itval);
+				power.x = fabs(itval);
+			}
+
+			// total desired speed along the axes
+			// note: "main cruise axls" is applied via cruise speed elsewhere
+			lin += strafe;
+		} else {
+			// no linear thrust key down
+			power.x = power.y = power.z = linearThrustPower;
+		}
+
+		// thrust-independent setting of cruise speed
+		if (!m_stickySpeedKey && InputBindings.speedControl->IsActive()) {
+			Util::ChangeCruiseSpeedConstrained(*this, InputBindings.speedControl->GetValue() * std::max(std::abs(m_cruiseSpeed) * 0.05, 1.0));
+		}
+
+		// apply speed limiter to cruise speed
+		if (m_speedLimiterActive && fabs(m_cruiseSpeed) > m_speedLimit) {
+			m_cruiseSpeed = m_speedLimit * Util::sign(m_cruiseSpeed);
+		}
+
+		// apply cruise speed
+		switch (GetCruiseDirection()) {
+		case CRUISE_FWD:
+			lin += -m_ship->GetOrient().VectorZ() * m_cruiseSpeed;
+			break;
+		case CRUISE_UP:
+			lin += m_ship->GetOrient().VectorY() * m_cruiseSpeed;
+			break;
+		default: assert(false && "Unknown cruise direction");
+		}
+
+		// if 'cruise speed' is stuck at 0, here we will free it
+		if (m_stickySpeedKey && !InputBindings.speedControl->IsActive()) {
+			if (GetCruiseDirection() == CRUISE_FWD) {
+				if (!InputBindings.thrustForward->IsActive()) {
+					m_stickySpeedKey = false;
+				}
+			} else { // CRUISE_UP
+				if (!InputBindings.thrustUp->IsActive()) {
+					m_stickySpeedKey = false;
+				}
+			}
+		}
+
+		// recalculate power limits, depending on external and inertial forces
+		// because we want to use all available thrust to counter these forces
+		// regardless of the set power limit
+		// the speed that these forces will develop over the upcoming frame
+		vector3d extfv = extf * timeStep / m_ship->GetMass();
+		// predicted direction of change in speed (and burning thrusters)
+		vector3d dvDir = (lin - m_ship->GetVelocity()) * m_ship->GetOrient() - extfv;
+		vector3d maxThrust = m_ship->GetPropulsion()->GetThrust(dvDir); // Newtons, positive only
+		// correction can only be in a greater value
+		vector3d comp = vector3d(
+			std::min(extf.x / maxThrust.x * Util::sign(dvDir.x), 0.0),
+			std::min(extf.y / maxThrust.y * Util::sign(dvDir.y), 0.0),
+			std::min(extf.z / maxThrust.z * Util::sign(dvDir.z), 0.0)); // fraction of thrust
+		power.x = Clamp(power.x - comp.x, 0.0, 1.0);
+		power.y = Clamp(power.y - comp.y, 0.0, 1.0);
+		power.z = Clamp(power.z - comp.z, 0.0, 1.0);
+
+	} else if (m_flightControlState == CONTROL_MANUAL) {
+	}
+}
+
+void PlayerShipController::ApplyTotalAction(const TotalDesiredAction &params)
+{
+	// setting the desired speed has priority over the low-level thrust setting
+	if (params.velLinear) {
+		m_ship->AIMatchVel(*params.velLinear, params.linPower);
+	} else if (params.thrLinear) {
+		m_ship->GetPropulsion()->SetLinThrusterState(*params.thrLinear);
+	} else {
+		m_ship->ClearThrusterState();
+	}
```

Comment:
If the TotalDesiredAction is rewritten as above, this block becomes:

```suggestion
	// setting the desired speed has priority over the low-level thrust setting
	if (params.desireLinVel) {
		m_ship->AIMatchVel(params.desiredLinear, params.linPower);
	} else {
		m_ship->GetPropulsion()->SetLinThrusterState(params.desiredLinear);
	}
```

Reply:
@Web-eWorks when to call `ClearThrusterState`? this is not exactly the same as `SetLinThrusterState({})`.

Reply:
ClearThrusterState has the exact same effect as calling SetLinThrusterState({0, 0, 0}), it's just a comparative "fast path" due to not needing to clamp and scale the thrust values. Given that this is called ~once per frame, I don't see any issues with taking the slow path here, and it can be optimized inside of SetLinThrusterState.

Reply:
@Web-eWorks not exactly - this is done so that the ship maintains thrust during takeoff for some time
https://github.com/pioneerspacesim/pioneer/blob/6ca9518f400619704c3fea49a4abb0d76b69c6a3/src/Ship.h#L91-L95

Reply:
...so *that's* where the launch thrust is handled. That really should be moved to an explicit routine somewhere instead of stuffed into `ClearThrusterState`, as this is one of the jankier hacks in the flight control system in my opinion. (For example, apply some forward thrust during the launch timeout - it should clear that forward thrust and continue launching upwards.)

Reply:
@Web-eWorks could you please describe in more detail how to implement this?

Reply:
@Web-eWorks Oh, are you suggesting storing either thrust or velocity in the same variable? Does it bother you that these are different units of measurement?

Reply:
The different units of measurement are mutually exclusive as far as I can tell - linear and angular motion are individually either used in match-velocity or total-thrust mode but never as thrust and match-velocity at the same time. Reusing the same variable for both modes makes sense, as the actual representation of the value does not change even if the interpretation does.

Reply:
@Web-eWorks yes, it will work, although now you will have to either check or assume the "type" of the data at the current point of execution each time you access this fields.

Reply:
As far as I was able to see looking through the relevant code, most functions using this datastructure are already implicitly assuming the entire datastructure is being used for one or the other, as the way the optionals are evaluated to set the thruster state precludes any use of the datastructure for both match-vel and total-thrust modes for the same component at the same time.

Reply:
Done, it's still nice that we managed to get rid of optional!